huggingface_pipelines/text.py

```diff
@@ -0,0 +1,156 @@
+        """Decodes a list of embeddings back into texts."""
+        try:
+            logger.info(f"Decoding {len(embeddings)} embeddings...")
+            decoded_texts = self.t2t_model.predict(
+                embeddings, target_lang=self.target_lang, batch_size=self.batch_size)
+            logger.info("Texts decoded successfully.")
+            return decoded_texts
+        except Exception as e:
+            logger.error(f"Error decoding texts: {e}")
+            return []
+
+    def compute_bleu(self, original_texts: List[str], reconstructed_texts: List[str]) -> List[float]:
+        """Computes the BLEU score between original and reconstructed texts."""
+        logger.info("Computing BLEU score...")
+
+        logger.info(f"Original texts: {original_texts}")
+        logger.info(f"Reconstructed Texts: {reconstructed_texts}")
+
+        references = [text.split()
+                      for text in original_texts]
+
+        logger.info(f"References: {references}")
+
+        predictions = reconstructed_texts
+
+        logger.info(f"Predictions: {predictions}")
+
+        bleu_score = self.bleu_metric.compute(
+            predictions=predictions, references=references, smooth=True)
+        logger.info(f"BLEU score computed: {bleu_score['bleu']}")
+        return [bleu_score['bleu']]
+
+    def process_batch(self, batch: Dict[str, Any]) -> Dict[str, Any]:
+        """Processes a single batch of data, returning original, reconstructed texts and BLEU score."""
+        logger.info("Processing batch...")
+        texts = batch['text']
+        embeddings = self.encode_texts(texts)
+        reconstructed_texts = self.decode_embeddings(embeddings)
+        bleu_score = self.compute_bleu(texts, reconstructed_texts)
+        logger.info(f"Batch processed. BLEU score: {bleu_score}")
+        return {'original': texts, 'reconstructed': reconstructed_texts, 'bleu': bleu_score}
+
+    def process_batches(self):
+        """Processes all batches in the dataset and stores the results."""
+        try:
+            logger.info("Starting to process batches...")
+            # Limit to the first 1000 examples
+            dataset = self.dataset.select(range(min(len(self.dataset), 1000)))
+
+            # Process the dataset using map
+            results = dataset.map(
+                lambda batch: self.process_batch(batch),
+                batched=True,
+                batch_size=self.batch_size,
+                remove_columns=dataset.column_names,
+                load_from_cache_file=False
+            )
+            self.results.extend([{k: v[i] for k, v in results.items()}
+                                for i in range(len(results[next(iter(results))]))])
+
+            logger.info("All batches processed. Caching results...")
+            self.cache_results()
+            logger.info("Results cached successfully.")
+        except Exception as e:
+            logger.error(f"Error processing batches: {e}")
+
+    def cache_results(self):
```

Comment:
let's cache result as arrow (or at least introduce an option for it).

Reply:
and I think we want to use already existing primitives from HF : https://huggingface.co/docs/datasets/en/about_cache